File: Kalorie/Kalorie/Core/UseCases/FetchMySubmissionsUseCase.swift
//
//  FetchMySubmissionsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation

protocol FetchMySubmissionsUseCaseProtocol {
    func callAsFunction() async throws -> [FoodItemSubmissionDomain]
}

struct FetchMySubmissionsUseCase: FetchMySubmissionsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction() async throws -> [FoodItemSubmissionDomain] {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        return try await FoodItemSubmissionFetcher.fetch(where: "submitted_by", isEqualTo: userId, dataProvider: dataProvider)
    }
}

#if DEBUG
struct FetchMySubmissionsUseCaseFake: FetchMySubmissionsUseCaseProtocol {

    // MARK: - Properties

    var stubbedSubmissions: [FoodItemSubmissionDomain] = []
    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction() async throws -> [FoodItemSubmissionDomain] {
        if shouldThrow { throw URLError(.unknown) }
        return stubbedSubmissions
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchMySubmissionsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemSubmissionDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol

interface FetchMySubmissionsUseCaseProtocol {
    suspend operator fun invoke(): List<FoodItemSubmissionDomain>
}

class FetchMySubmissionsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : FetchMySubmissionsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(): List<FoodItemSubmissionDomain> {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        return FoodItemSubmissionFetcher.fetch(field = "submitted_by", isEqualTo = userId, dataProvider = dataProvider)
    }
}
